testing/integration/otel_test.go

```diff
@@ -1099,3 +1100,191 @@ service:
+		if t.Failed() {
+			contents, err := os.ReadFile(otelConfigPath)
+			if err != nil {
+				t.Logf("No otel configuration file at %s", otelConfigPath)
+				return
+			}
+			t.Logf("Contents of otel config file:\n%s\n", string(contents))
+		}
+	})
+	// Now we can actually create the fixture and run it
+	fixture, err := define.NewFixtureFromLocalBuild(t, define.Version(), aTesting.WithAdditionalArgs([]string{"--config", otelConfigPath}))
+	require.NoError(t, err)
+
+	ctx, cancel := testcontext.WithDeadline(t, context.Background(), time.Now().Add(5*time.Minute))
+	defer cancel()
+	err = fixture.Prepare(ctx, fakeComponent)
+	require.NoError(t, err)
+
+	// Write logs to input file
+	var inputLinesCounter atomic.Int64
+	var stopInputWriter atomic.Bool
+	go func() {
+		for i := 0; ; i++ {
+			if stopInputWriter.Load() {
+				break
+			}
+
+			_, err = inputFile.Write([]byte(fmt.Sprintf("Line %d\n", i)))
+			require.NoErrorf(t, err, "failed to write line %d to temp file", i)
+			inputLinesCounter.Add(1)
+			time.Sleep(100 * time.Millisecond)
+		}
+		err = inputFile.Close()
+		require.NoError(t, err, "failed to close input file")
+	}()
+
+	t.Cleanup(func() {
+		if t.Failed() {
+			contents, err := os.ReadFile(inputFilePath)
+			if err != nil {
+				t.Logf("no data file to import at %s", inputFilePath)
+				return
+			}
+			t.Logf("contents of import file:\n%s\n", string(contents))
+		}
+	})
+
+	var restartCount atomic.Int64
+	var fixtureWg sync.WaitGroup
+	fixtureWg.Add(1)
+	go func() {
+		defer fixtureWg.Done()
+
+		for {
+			// restart the collector every couple seconds while new data is being written
+			prevCount := inputLinesCounter.Load()
+			fCtx, cancel := context.WithDeadline(ctx, time.Now().Add(2*time.Second))
+			err = fixture.RunOtelWithClient(fCtx)
+			cancel()
+			require.True(t, errors.Is(err, context.DeadlineExceeded) || errors.Is(err, context.Canceled), "unexpected error: %v", err)
+			require.True(t, inputLinesCounter.Load() > prevCount, "expected input lines to increase")
+			restartCount.Add(1)
+
+			if stopInputWriter.Load() {
+				break
+			}
+		}
+
+		// start the collector again for the remaining of the test
+		fCtx, cancel := context.WithDeadline(ctx, time.Now().Add(5*time.Minute))
+		defer cancel()
+		err = fixture.RunOtelWithClient(fCtx)
+	}()
+
+	time.Sleep(5 * time.Second)
+	stopInputWriter.Store(true)
+
+	require.True(t, restartCount.Load() > 0, "expected the collector to restart at least once")
+
+	// Make sure all the logs are ingested
+	actualHits := &struct{ Hits int }{}
+	require.Eventually(t,
+		func() bool {
+			findCtx, findCancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer findCancel()
+
+			docs, err := estools.GetLogsForIndexWithContext(findCtx, info.ESClient, ".ds-"+index+"*", map[string]interface{}{
+				"log.file.path": inputFilePath,
+			})
+			require.NoError(t, err)
+
+			actualHits.Hits = docs.Hits.Total.Value
+			return actualHits.Hits == int(inputLinesCounter.Load())
```

Comment:
This doesn't guarantee we read each unique log line, only that we read an amount of log lines equal to the number that was written, regardless of if some of them are duplicates?

The test we need is "every log line in the file was shipped at least once" and I don't think this is that.

If you gave the log lines you are shipping an _id field Elasticsearch would deduplicate them for us and then the count would be enough. There is a blog post explaining this process and some sample configurations for Beats and Elastic Agent https://www.elastic.co/blog/log-deduplication-with-elasticsearch